src/main/java/com/codacy/client/stash/client/auth/OAuthSignatureCalculatorRSA.java

```diff
@@ -0,0 +1,347 @@
+        return System.currentTimeMillis() / 1000L;
+    }
+
+    protected String generateNonce() {
+        byte[] nonceBuffer = NONCE_BUFFER.get();
+        ThreadLocalRandom.current().nextBytes(nonceBuffer);
+        // let's use base64 encoding over hex, slightly more compact than hex or decimals
+        return Base64.encode(nonceBuffer);
+//      return String.valueOf(Math.abs(random.nextLong()));
+    }
+
+    /**
+     * Container for parameters used for calculating OAuth signature.
+     * About the only confusing aspect is that of whether entries are to be sorted
+     * before encoded or vice versa: if my reading is correct, encoding is to occur
+     * first, then sorting; although this should rarely matter (since sorting is primary
+     * by key, which usually has nothing to encode)... of course, rarely means that
+     * when it would occur it'd be harder to track down.
+     */
+    final static class OAuthParameterSet {
+        private final ArrayList<Parameter> allParameters;
+
+        public OAuthParameterSet(int size) {
+            allParameters = new ArrayList<>(size);
+        }
+
+        public OAuthParameterSet add(String key, String value) {
+            Parameter p = new Parameter(UTF8UrlEncoder.encode(key), UTF8UrlEncoder.encode(value));
+            allParameters.add(p);
+            return this;
+        }
+
+        public String sortAndConcat() {
+            // then sort them (AFTER encoding, important)
+            Parameter[] params = allParameters.toArray(new Parameter[allParameters.size()]);
+            Arrays.sort(params);
+
+            // and build parameter section using pre-encoded pieces:
+            StringBuilder encodedParams = new StringBuilder(100);
+            for (Parameter param : params) {
+                if (encodedParams.length() > 0) {
+                    encodedParams.append('&');
+                }
+                encodedParams.append(param.key()).append('=').append(param.value());
+            }
+            return encodedParams.toString();
+        }
+    }
+
+    /**
+     * Helper class for sorting query and form parameters that we need
+     */
+    final static class Parameter implements Comparable<Parameter> {
+        private final String key, value;
+
+        public Parameter(String key, String value) {
+            this.key = key;
+            this.value = value;
+        }
+
+        public String key() {
+            return key;
+        }
+
+        public String value() {
+            return value;
+        }
+
+        @Override
+        public int compareTo(Parameter other) {
+            int diff = key.compareTo(other.key);
+            if (diff == 0) {
+                diff = value.compareTo(other.value);
+            }
+            return diff;
+        }
+
+        @Override
+        public String toString() {
+            return key + "=" + value;
+        }
+
+        @Override
+        public boolean equals(Object o) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [The method equals() has an NPath complexity of 24](https://www.codacy.com/app/Codacy/stash-scala-client/file/2157129301/issues/source?bid=2734487&fileBranchId=3159563#l328)